lib/semantics/check-do.cc

```diff
@@ -647,20 +645,131 @@ void DoChecker::Leave(const parser::DoConstruct &x) {
+  }
+  return nullptr;
+}
+
+// Check that CYCLE and EXIT statements do not cause flow of control to
+// leave DO CONCURRENT, CRITICAL, or CHANGE TEAM constructs.
+void DoChecker::CheckForBadLeave(
+    const char *stmtName, const ConstructNode &construct) const {
+  // C1135 and C1167
+  if (const auto doConstructPtr{MaybeGetDoConstruct(construct)}) {
+    if (doConstructPtr->IsDoConcurrent()) {
+      SayBadLeave(stmtName, "DO CONCURRENT", construct);
+    }
+    return;
+  }
+  // C1135 and C1168
+  if (const auto criticalConstructPtr{
+          std::get_if<const parser::CriticalConstruct *>(&construct)}) {
+    SayBadLeave(stmtName, "CRITICAL", construct);
+    return;
+  }
+  if (const auto changeTeamConstructPtr{
+          std::get_if<const parser::ChangeTeamConstruct *>(&construct)}) {
+    SayBadLeave(stmtName, "CHANGE TEAM", construct);
+    return;
+  }
+}
+
+static bool ConstructIsDoConcurrent(const ConstructNode &construct) {
+  parser::DoConstruct const *doConstruct{MaybeGetDoConstruct(construct)};
+  return doConstruct && doConstruct->IsDoConcurrent();
+}
+
+static bool isExitStmt(const char *stmtType) {
+  return std::strncmp("EXIT", stmtType, 4) == 0;
+}
+
+static bool StmtMatchesConstruct(NamePtr stmtName, const char *stmtType,
+    NamePtr constructName, const ConstructNode &construct) {
+  bool inDoConstruct{MaybeGetDoConstruct(construct) != nullptr};
+  if (stmtName == nullptr) {
+    if (inDoConstruct) {
+      return true;  // Unlabeled statements match all DO constructs
+    } else {
+      return false;  // Unlabeled statements match no non-DO constructs
+    }
+  } else if (!constructName) {
+    return false;  // name on CYCLE/EXIT, but not on the construct
+  } else if (constructName->source == stmtName->source) {
+    if (isExitStmt(stmtType)) {
+      return true;  // EXIT name matches any construct name
+    } else {
+      if (inDoConstruct) {
+        return true;  // CYCLE name matches only DO construct name
+      }
+    }
+  }
+  return false;
+}
+
+// C1167 Can't EXIT from a DO CONCURRENT
+void DoChecker::CheckDoConcurrentExit(
+    const char *stmtType, const ConstructNode &construct) const {
+  if (isExitStmt(stmtType) && ConstructIsDoConcurrent(construct)) {
+    SayBadLeave("EXIT", "DO CONCURRENT", construct);
   }
 }
 
-// C1166
-void DoChecker::Enter(const parser::ExitStmt &) {
-  if (!context_.InsideDoConstruct()) {
-    context_.Say(
-        *context_.location(), "EXIT must be within a DO construct"_err_en_US);
+// Check nesting violations for a CYCLE or EXIT statement Loop up the nesting
+// levels looking for a construct that matches the CYCLE or EXIT statment.  At
+// every construct, check for a violation.  If we find a match without finding
+// a violation, the check is complete.
+void DoChecker::CheckNesting(const char *stmtType, NamePtr stmtName) const {
+  const ConstructStack &stack{context_.constructStack()};
+  for (ConstructStack::const_reverse_iterator riter = stack.crbegin();
```

Comment:
can you make this `rIter`?

Reply:
and use braces, and consider omitting the explicit type.

Reply:
All sounds good.  I'll change the code.